model/map/src/main/java/org/keycloak/models/map/singleUseObject/MapSingleUseObjectProvider.java

```diff
@@ -0,0 +1,234 @@
+        if (Time.currentTime() < expiration)  {
+            return new MapSingleUseObjectAdapter(session, origEntity);
+        } else {
+            actionTokenStoreTx.delete(origEntity.getId());
+            return null;
+        }
+    }
+
+    @Override
+    public void put(ActionTokenKeyModel actionTokenKey, Map<String, String> notes) {
+        if (actionTokenKey == null || actionTokenKey.getUserId() == null || actionTokenKey.getActionId() == null || actionTokenKey.getActionVerificationNonce() == null) {
+            return;
+        }
+
+        LOG.tracef("put(%s, %s, %s)%s", actionTokenKey.getUserId(), actionTokenKey.getActionId(), actionTokenKey.getActionVerificationNonce(), getShortStackTrace());
+
+        DefaultModelCriteria<ActionTokenValueModel> mcb = criteria();
+        mcb = mcb.compare(ActionTokenValueModel.SearchableFields.USER_ID, ModelCriteriaBuilder.Operator.EQ, actionTokenKey.getUserId())
+                .compare(ActionTokenValueModel.SearchableFields.ACTION_ID, ModelCriteriaBuilder.Operator.EQ, actionTokenKey.getActionId())
+                .compare(ActionTokenValueModel.SearchableFields.ACTION_VERIFICATION_NONCE, ModelCriteriaBuilder.Operator.EQ, actionTokenKey.getActionVerificationNonce().toString());
+
+        ActionTokenValueModel existing = actionTokenStoreTx.read(withCriteria(mcb))
+                .findFirst().map(this::singleUseEntityToAdapter).orElse(null);
+
+        if (existing == null) {
+            MapSingleUseObjectEntity actionTokenStoreEntity = new MapSingleUseObjectEntityImpl();
+            actionTokenStoreEntity.setUserId(actionTokenKey.getUserId());
+            actionTokenStoreEntity.setActionId(actionTokenKey.getActionId());
+            actionTokenStoreEntity.setActionVerificationNonce(actionTokenKey.getActionVerificationNonce().toString());
+            actionTokenStoreEntity.setExpiration(TimeAdapter.fromIntegerWithTimeInSecondsToLongWithTimeAsInSeconds(actionTokenKey.getExpiration()));
+            actionTokenStoreEntity.setNotes(notes);
+
+            LOG.debugf("Adding used action token to actionTokens cache: %s", actionTokenKey.toString());
+
+            actionTokenStoreTx.create(actionTokenStoreEntity);
+        }
+    }
+
+    @Override
+    public ActionTokenValueModel get(ActionTokenKeyModel key) {
+        LOG.tracef("get(%s, %s, %s)%s", key.getUserId(), key.getActionId(), key.getActionVerificationNonce(), getShortStackTrace());
+
+        DefaultModelCriteria<ActionTokenValueModel> mcb = criteria();
+        mcb = mcb.compare(ActionTokenValueModel.SearchableFields.USER_ID, ModelCriteriaBuilder.Operator.EQ, key.getUserId())
+                .compare(ActionTokenValueModel.SearchableFields.ACTION_ID, ModelCriteriaBuilder.Operator.EQ, key.getActionId())
+                .compare(ActionTokenValueModel.SearchableFields.ACTION_VERIFICATION_NONCE, ModelCriteriaBuilder.Operator.EQ, key.getActionVerificationNonce().toString());
```

Comment:
We might want to add similar check here as well, right? There are also other methods within the class.

Reply:
I added the guard to the `get(...)` and `remove(...)` methods.